rust/cargo_manifest_to_build.py

```diff
@@ -0,0 +1,561 @@
+    crate_root = "{path}",
+    crate_type = "{kind}",
+    srcs = glob(["**/*.rs"]),
+    deps = [{deps}],
+    rustc_flags = [
+        "--cap-lints allow",{flags}
+    ],{other}
+    version = "{version}",
+    crate_features = [{features}],
+)""".format(
+    name=target["name"].replace("-", "_"),
+    path=remove_path_prefix(target["src_path"], context["workspace_root"]),
+    kind=target["crate_types"][0],
+    deps=context["deps"],
+    flags="".join(["\n        \"%s\"," % f for f in context["flags"]]),
+    other=context["other"],
+    version=context["version"],
+    features=",".join("\"%s\"" % f for f in context["features"]),
+)
+
+def rust_binary(target, context):
+    return """rust_binary(
+    name = "{name}_bin",
+    crate_root = "{path}",
+    srcs = glob(["**/*.rs"]),
+    deps = [":{name}", {deps}],
+    rustc_flags = [
+        "--cap-lints allow",{flags}
+    ],{other}
+    version = "{version}",
+    crate_features = [{features}],
+)""".format(
+    name=target["name"].replace("-", "_"),
+    path=remove_path_prefix(target["src_path"], context["workspace_root"]),
+    deps=context["deps"],
+    flags="".join(["\n        \"%s\"," % f for f in context["flags"]]),
+    other=context["other"],
+    version=context["version"],
+    features=",".join("\"%s\"" % f for f in context["features"]),
+)
+
+def custom_build_script(target, context):
+    return """rust_binary(
+    name = "{name}_build_script",
+    crate_root = "{path}",
+    srcs = glob(["**/*.rs"]),
+    deps = [{build_deps}],
+    rustc_flags = ["--cap-lints allow"],
+    version = "{version}",
+    crate_features = [{features}],
+    visibility = ["//visibility:private"],
+)
+
+cargo_build_script_run(
+    name = "{name}_build_script_executor",
+    srcs = glob(["*", "**/*.rs"]),
+    script = ":{name}_build_script",
+    features = [{features}],
+)
+""".format(
+    name=context["name"].replace("-", "_"),
+    path=remove_path_prefix(target["src_path"] or "build.rs", context["workspace_root"]),
+    build_deps=context["build_deps"],
+    features=",".join("\"%s\"" % f for f in context["features"]),
+    version=context["version"],
+)
+
+def resolve_deps(deps, ctxt):
+    return ",".join([
+        "\"%s\"" % ctxt["dep_format"].format(
+            name=d["name"].replace("-", "_"),
+            version=ctxt["resolved_deps"][d["name"]].replace(".", "_")
+        )
+        for d in deps
+        if d["name"] in ctxt["resolved_deps"]
+        ])
+
+def extend_context(ctxt, json):
+    out_dir_tar = ""
+    if any([t["kind"] == ["custom-build"]] for t in json["targets"]):
+        out_dir_tar = "\n    out_dir_tar = \":%s_build_script_executor\"," % json["name"].replace("-", "_")
+    data = ""
+    if ctxt["data"]:
+        data = "\n    data = r\"\"\"%s\"\"\"," % ctxt["data"]
+    for d in json["dependencies"]:
+        if d["name"] not in ctxt["resolved_deps"]:
+            sys.stderr.write("WARNING: Cannot resolve dependency on crate '%s'\n" % d["name"])
```

Comment:
This isn't  necessarily erroneous.

If cargo-raze determines that the given combination of `features` and `target` indicate that a feature should be omitted, then it will not be present in the set of resolved deps.

Reply:
Right maybe we should add a "n/a" value that explicitely says we should skip that deps. That way people that right themselves the target will get the warning if they omit a dep?